### PR TITLE
Add C++ Solution + Editorial to The Great Revegetation (Silver)

### PR DESCRIPTION
**Pull request by ryanchou-dev**
yayy

---
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/usaco-920.mdx [hunk 1]
```diff
@@ -2,13 +2,102 @@
...
-author: Sofia Yang
+author: Sofia Yang, Ryan Chou
 ---
 
-[Official Analysis](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)
 
+Assume that we have a valid solution composed of $K$ connected components in our graph.
+
+Now, let's look at one of these, which could look something like this:
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:12]:
```suggestion
Let's look at one of these connected components, which could look something like this:
```

File: solutions/usaco-920.mdx [hunk 2]
```diff
@@ -2,13 +2,102 @@
...
+
+![possible connected component](<usaco-920/connected_component.png>)
+
+However, for all $K$ connected component, the grass types can be "inverted":
+
+![inverted connected component](<usaco-920/inverted_connected_component.png>)
+
+So, for each $K$ connected components, there are $2$ possible ways to arrange them, or $2^k$ ways to arrange the entire graph. Since in binary, $2^k$ is $1$ followed by $k$ zeros, we can simply print this instead of manually converting it.
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:20]: how about
"for each connected component in our graph..."
"thus, if there's k connected components, there's 2^k ways to arrange"

File: solutions/usaco-920.mdx [hunk 3]
```diff
@@ -2,13 +2,102 @@
...
+					}
+				}
+			}
+		}
+	}
+
+
+	if (impossible) {
+		cout << 0 << "\n";
+		return 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:84]: no need for return 0

**Review by nathangong — COMMENTED** [review]
same thoughts as sans

**Comment by ryanchou-dev** [comment]
thanks for reviewing!

File: solutions/usaco-920.mdx [hunk 4]
```diff
@@ -2,13 +2,102 @@
...
-[Official Analysis](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)
 
+Assume that we have a valid solution composed of $K$ connected components in our graph.
+
+Now, let's look at one of these, which could look something like this:
+
+![possible connected component](<usaco-920/connected_component.png>)
+
+However, for all $K$ connected component, the grass types can be "inverted":
```
[the remark this answers is not in the document]
Comment (ryanchou-dev) [reply solutions/usaco-920.mdx]: oops, yeah now that i read this, it sounds really weird

File: solutions/usaco-920.mdx [hunk 5]
```diff
@@ -2,13 +2,101 @@
...
-[Official Analysis](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)
 
+Assume that we have a valid solution composed of $K$ connected components in our graph.
+
+Let's look at one of these connected components, which could look something like this:
+
+![possible connected component](<usaco-920/connected_component.png>)
+
+However, for all $K$ connected component, the grass types can be flipped and still be a valid solution:
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:16]: > for all $K$ connected component

component**s**

File: solutions/usaco-920.mdx [hunk 6]
```diff
@@ -2,13 +2,101 @@
...
+
+	if (impossible) {
+		cout << 0 << endl;
+	} else {
+		/*
+		 * if we can construct our solution, then there are 2^cc possible rearrangements,
+		 * since we can always invert the types of grass in each connected component.
+		 */
+
+		// 2^cc in binary is 1, followed by cc zeros.
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:90]: update this comment

File: solutions/usaco-920.mdx [hunk 7]
```diff
@@ -2,13 +2,101 @@
...
+	vector<int> color(n, -1);
+
+	for (int i = 0; i < n; i++) {
+		if (color[i] == -1) {
+			component_num++;
+			
+			queue<pair<int, bool>> todo;
+			todo.push({i, true});
+
+			while ((int)(todo.size())) {
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:59]: `!todo.empty()` instead

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Review by SansPapyrus683 — APPROVED** [review]

**Review by TruaShamu — APPROVED** [review]

**Comment by TruaShamu** [comment]
Also, this is optional but you can use these diagrams if you want:

connected_component
inverted_connected_component

**Comment by ryanchou-dev** [comment]
> Also, this is optional but you can use these diagrams if you want:
> 
> connected_component inverted_connected_component

omg this looks so much better.

thanks!

File: solutions/usaco-920.mdx [hunk 8]
```diff
@@ -2,13 +2,101 @@
...
+
+
+	if (impossible) {
+		cout << 0 << endl;
+	} else {
+		/*
+		 * if we can construct our solution, then there are 2^cc possible rearrangements,
+		 * since we can always invert the types of grass in each connected component.
+		 */
+
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:89]:
```suggestion
```
alr thoroughly explained in editorial

File: solutions/usaco-920.mdx [hunk 9]
```diff
@@ -2,13 +2,101 @@
...
+			todo.push({i, true});
+
+			while (!todo.empty()) {
+				pair<int, bool> nxt = todo.front();
+				todo.pop();
+
+				color[nxt.first] = nxt.second;
+
+				for (pair<int, bool> u : adj[nxt.first]) {
+					bool type = (u.second ? nxt.second : !nxt.second);
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:66]:
```suggestion
					bool type = u.second ? nxt.second : !nxt.second;
```

File: solutions/usaco-920.mdx [hunk 10]
```diff
@@ -2,13 +2,101 @@
...
+		adj[--u].push_back({--v, type == 'S'});
+		adj[v].push_back({u, type == 'S'});
+	}
 
+	int component_num = 0;
+	bool impossible = false;
+	vector<int> color(n, -1);
+
+	for (int i = 0; i < n; i++) {
+		if (color[i] == -1) {
```
Comment (SansPapyrus683) [inline solutions/usaco-920.mdx:53]: comments for this entire part

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
when build passes i'll merge

**Pull request merged**